lib/fluent/plugin/in_sample.rb

```diff
@@ -117,15 +119,23 @@ def emit(num)
       end
     end
 
+    def sample
+      if @reuse_record
+        @sample[@sample_index]
+      else
+        @sample[@sample_index].dup
+      end
+    end
```

Comment:
It would be better if the responsibilities of methods were more divided.

In the current design, the `sample` method depends on `@sample_index`, but `@sample_index` is handled by the `genetate` method.
This appears to have unnecessarily increased the mutual dependence of these two methods.

There would be two possible ways to improve this.

* Way 1:  This method handles the index as well.

```rb
    def next_sample
      d = @reuse_record ? @sample[@sample_index] : @sample[@sample_index].dup
      @sample_index += 1
      return d if d

      @sample_index = 0
      next_sample
    end
```

* Way 2: This method does not depend on `@sample_index`.

```rb
    def sample(index)
      if @reuse_record
        @sample[index]
      else
        @sample[index].dup
      end
    end
```